### PR TITLE
Guide for training on your own data

### PR DESCRIPTION
**Pull request by dlwh**

### CODE REVIEW

File: docs/Training-On-Your-Data.md [hunk 1]
```diff
@@ -0,0 +1,354 @@
...
+TODO: add instructions for llama including auth.
+
+You should probably reduce the learning rate by a factor of 10 or so. TODO: figure out best practices here.
+
+
+### Checkpointing
+
+See also the [Checkpointing section of the Configuration Guide](./Configuration-Guide.md#checkpointing).
+
+Levanter supports checkpointing to both local and Google Cloud Storage, backed by [Tensorstore](https://google.github.io/tensorstore/).
```
Comment (Ivan-Zhou) [inline docs/Training-On-Your-Data.md:218]:
```suggestion
Levanter supports checkpointing to both local and Google Cloud Storage, backed by [TensorStore](https://google.github.io/tensorstore/).
```

File: docs/Training-On-Your-Data.md [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+
+You can actually connect this to a Ray cluster, and use the cluster to do the preprocessing. This lets you
+use any of Ray's autoscaling features to scale up the preprocessing job.
+
+To do so:
+
+```bash
+python -m levanter.main.cache_dataset \
+    --config_path my_config.yaml \
+    --address <coordinator-address> \
```
Comment (Ivan-Zhou) [inline docs/Training-On-Your-Data.md:134]:
```suggestion
    --address <ray-cluster-address> \
```

Maybe this is clearer, but not sure

File: docs/Training-On-Your-Data.md [hunk 3]
```diff
@@ -0,0 +1,354 @@
...
+Levanter supports starting from an HF pretrained model. To do so, you should set your config like this:
+
+```yaml
+model:
+  type: mpt
+initialize_from_hf: "mosaicml/mpt-7b@68e1a8e0ebb9b30f3c45c1ef6195980f29063ae2"
+use_hf_model_config: true
+```
+
+TODO: add instructions for llama including auth.
```
Comment (Ivan-Zhou) [inline docs/Training-On-Your-Data.md:209]: In my case, I just pass my token as env variable. Would it be possible to set in the yaml?

```
WANDB_API_KEY=$WANDB_API_KEY HUGGING_FACE_HUB_TOKEN=$HF_TOKEN  levanter/infra/run.sh python3 ...
```

File: docs/Training-On-Your-Data.md [hunk 4]
```diff
@@ -0,0 +1,354 @@
...
+your job runs. Note that, due to FSDP, as you add more TPUs, you can increase the effective parallelism because
+you will use less memory per accelerator to store parameters and optimizer states.
+
+### Number of Training Steps
+
+Levanter does not support epochs or number of tokens/examples, so if you want to train for a certain number of epochs or
+tokens, you'll need to compute the number of steps yourself. You can use the following formula:
+
+```
+num_train_steps = num_epochs * num_tokens_per_epoch / train_batch_size / seq_len
```
Comment (Ivan-Zhou) [inline docs/Training-On-Your-Data.md:252]: If # tokens in an example is larger than seq_len, do we break it into multiple examples?

File: docs/Training-On-Your-Data.md [hunk 5]
```diff
@@ -0,0 +1,354 @@
...
+    WANDB_API_KEY=... levanter/infra/run.sh python levanter/src/levanter/main/train_lm.py --config_path gs://path/to/config.yaml
+```
+
+#### Spin up and manual launch
+
+You should probably use the automated setup script, as described in the [relevant section of the TPU guide](./Getting-Started-TPU-VM.md#automatic-setup).
+Here's what that looks like:
+
+```bash
+bash infra/spin-up-tpu-vm.sh my-tpu -z us-east1-d -t v3-128
```
Comment (Ivan-Zhou) [inline docs/Training-On-Your-Data.md:292]:
```suggestion
bash infra/spin-up-vm.sh my-tpu -z us-east1-d -t v3-128
```

Also, does this run setup? I thought we will need to attach the set up script, like:

```
bash infra/spin-up-vm.sh $NAME -b $BRANCH -z $ZONE -t $TYPE -s infra/setup-tpu-vm.sh
```

File: docs/Training-On-Your-Data.md [hunk 6]
```diff
@@ -0,0 +1,354 @@
...
+
+### Online Preprocessing
+
+We have a sophisticated caching mechanism using [Ray](https://docs.ray.io/en/latest/)
+that builds a cache of preprocessed data on the fly. Online caching happens transparently
+in the background, using the mostly-idle CPU-cores of the machine(s) you are training on.
+
+The cache that is built is fully reproducible, and can be used for future training runs.
+Training will start as soon as each training machine has its first shard of data cached
+and once the validation data is cached.
```
[the remark this answers is not in the document]
Comment (dlwh) [reply docs/Training-On-Your-Data.md:113]: Eh, I'm inclined to leave that out of this guide. I'll add a link to the sidebar.

[the remark this answers is not in the document]
Comment (dlwh) [reply docs/Training-On-Your-Data.md:113]: (my reasoning is the guide is pretty introductory and the design is pretty in-the-weeds

File: docs/Training-On-Your-Data.md [hunk 7]
```diff
@@ -0,0 +1,354 @@
...
+  hidden_dim: 1536
+  num_heads: 24
+  num_layers: 48
+  seq_len: 1024
+  gradient_checkpointing: true
+  scale_attn_by_inverse_layer_idx: true
+trainer:
+  wandb:
+    project: "levanter" # TODO
+    tags: [ "openwebtext", "gpt2"]
```
[the remark this answers is not in the document]
Comment (dlwh) [reply docs/Training-On-Your-Data.md:170]: oops, good call

File: docs/Training-On-Your-Data.md (hunk 3 above)
Comment (dlwh) [reply docs/Training-On-Your-Data.md:209]: yep! I just figured this out myself!

File: docs/Training-On-Your-Data.md (hunk 4 above)
Comment (dlwh) [reply docs/Training-On-Your-Data.md:252]: we do

**Review by Ivan-Zhou — APPROVED** [review]

**Pull request merged**